### ISSUE TITLE
[3.0] Standardise setting of gateway ihnterface language

### ISSUE THREAD

**Issue by judgej** [thephpleague/omnipay#279]
Many payment gateways have been based in single countries, and have only had to deal with a single language in their uder interface. Gateways have been increasingly expanding internationally over the last few years, and the interfaces have more of a need to cater for different languages.

It is a good time to standardise how OmniPay handles the selection of languages across the gateway drivers.

## Current Standards

ISO 639 defines codes for a list of languages. Each code consists of two lower-case letters. Examples are en, de, fr.

ISO 3166 defines codes for the countries of the world. The two-digit form of this standard consists of upper-case pairs of letters. Examples are GB, US, FR, AU.

Since many languages have variants that change across countries, it is convenient to define any given variation of a spoken or written language by its language code and its country of use. These are paired up into a format known as a locale. A locale tables the form of {language-code}-{country-code}. Examples are en-GB, en-US, fr-BE, fr-FR.

## Gateway Language Support

Typically, a gateway can have its interface language set through the API when making requests. Some gateways will fall back to the language of the user's browser if not specified, dropping back to the default language of the gateway if the browser language is not detected or supported.

The means and support to set the language varies across gateways, so we should standardise that in OmniPay. It makes sense being able to supply the locale to the gateway driver, and having that driver translate the locale to the supported languages of the gateway.

It should be noted that the documentation for many gateways is often outdated. so support for multiple languages is often better than the documentation suggests.

## Gateway Examples

Mollie accepts a field named "locale". The documentation provides some examples of non-standard locale values. However, the Mollie API does accept a fully-formed locale, and it is recommended that the language is requested in that form.

SagePay protocol v3 introduces a "language" field that takes the ISO 639 language code. The language code can be easily extracted from a locale.

MultiSafePay accepts a "locale" field in the form AB_cd (ISO 639 language code + ISO 3166 country code, with an underscore). I suspect the letter-case is incorrect in the documentation. But whatever the real format here, it can be derived from the ISO locale.

The PayPal OmniPay driver takes a "LocaleCode" field, which is passed to PayPal Express as field "LOCALECODE". Allowable values are listed here:

https://developer.paypal.com/docs/classic/express-checkout/integration-guide/ECCustomizing/

and they include some locale strings (language+country, with underscores) and some country codes alone.

TargetPay has a "lang" field that sets the user interface language. Values are nl, en, fr, de.

## Proposal

The "locale" field is defined and accepted by the omnipay-common abstract request. A locale value object is passed in. The locale value object can parse a locale string in a variety of formats (mixed letter-case, underscores/dashes, missing country code, etc.). Alternatively a string may be passed in, and that is parsed and converted to a value object when it is needed for a gateway message.

This does raise a question that I have come up against a number of times in the test scripts: should the parameter getter methods always return exactly the same as the parameter setting methods are given? If I `setLocale('en-GB')` is is okay for `getLocale()` to return a League\OmniPay\Interfaces\LocaleInterface object? The standard tests assume what comes out is always exactly what goes in. Does the setter/getter pattern insist that the setter and getter methods are totally dumb like that?

## locale String Format

A locale string can contain more than just a language and ounntry. So long as the language and country can be extracted, then it servies its purpose. I propose a locale object to handle all the parts of a locale that may be relevant. The PHP intl \Locale class will provide much of the functionality, such as parsing of locale strings, but cannot be instantiated, so is no use for passing locale values around.

**Comment by judgej** [comment]
Feedback, comments, questions, all expected :-) This is just a first draft of my thoughts on setting a gateway language (which are all done in various adhoc ways at the moment) so is just the start of a conversation.

**Comment by aimeos** [comment]
We absolutely support that proposal as we face the same problem in the Aimeos web shop components (http://aimeos.org)

For the locale string format we would suggest a simpler solution by only standardizing on either [POSIX format](https://en.wikipedia.org/wiki/Locale) (en_US) or [IETF / BCP 47 format](https://en.wikipedia.org/wiki/IETF_language_tag) (en-US[-...]) if the country is part of the locale string. IETF may be preferred if even regional locale parts should be supported. Otherwise and most often, it will be simply the language code (en).

Parsing locale strings should be only done by a regex, e.g. "([a-z]{2})(-[A-Z]{2})(-[A-Z]{3})" as it will save a lot of code and the dependency to the PHP intl extension. The result of an invalid code would be up to the payment gateway but that's OK.

**Comment by judgej** [comment]
Isn't the intl extension available for all versions of PHP from 5.3? I may be wrong, but if that is guaranteed then it means a lot of the tools will be available for parsing, constructing and extracting parts of locale strings in a number of formats.

Whatever the implementation, but wrapping that functionality into a locale object, with an interface to supply all the various parts of a locale, that is at least a chunk of the problem that the gateway drivers do not have to worry about; they are given a value object and are able to take out what they need, without the need to know anything about the original format.

**Comment by barryvdh** [comment]
I think it would be good to create a Locale object. The IETF format (`en-US`) is most used in HTML etc right? And `en_US` in PHP itself. But wouldn't be too hard to parse both in an object with the correct parts and let the gateway decide which format they need.

But do all gateways support all languages? Do we need to think about making this configurable? Eg. only set the language when 1 of some array (or fallback to a default). Or for 'exotic' formats, let the gateways create a mapper array/callback etc?

**Comment by judgej** [comment]
The mapper sounds good. If the locale class could provide both the language (en) and the localisation (US) as separate data, then that would provide some flexibility.

**Comment by aimeos** [comment]
I think this still sounds like an over-engineered solution for the problem. You only have "en", "en_US" or "en-US" as widely used possibilities. If we simply agree on one format, a getLocale()/setLocale('en_US') method would be enough. Please don't make it more complicated than it's really necessary.

**Comment by barryvdh** [comment]
We can only decide on the input format. The mapping/transforming is needed to create the correct value for the gateway itself.

**Comment by aimeos** [comment]
Yes, but that should be up to the driver if there's any transformation necessary.

**Comment by barryvdh** [comment]
Yes, but that's what I propose, we parse the input locale in parts, and just provide access to those parts, or perhaps 2/3 common formats. If they're different, gateways implement their own callback/mapper.

**Comment by aimeos** [comment]
Our proposal is even simpler: Standardize on one format and leave it as is, e.g.
``` Gateway::setLocale('en_US'); // or just "en"```

* Standardized locale: ```Gateway::getLocale(); // en_US```
* Language only: ```Gateway::getLocale(false); // en```
* IETF locale: ```str_replace( '_', '-', Gateway::getLocale() ); // en-US```

**Comment by barryvdh** [comment]
See https://github.com/thephpleague/omnipay/pull/349
